### PR TITLE
Add Three Alternative Models for Clear Sky Emissivity Calculation

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
There are two major algorithms adopted in BPS programs, one based on the work of Clark & Allen (1978), the other Martin & Berdahl (1984). Currently, EnergyPlus uses the Clark-Allen model based on a logarithmic relationship to dew point temperature, which was formulated using measurements limited to data collected over one year in San Antonio, Texas. While the original authors reported a low root mean square error (RMSE) of 10 W/m<sup>2</sup> while used to calculate longwave radiations, studies by the International Energy Agency and Dai and Fang demonstrated that, out of the empirical models examined, Clark-Allen had among the highest high errors when tested against MODTRAN predictions and observed data, respectively. Further illustrating the limitations of Clark-Allen, a more recent assessment by Zhang et al shows that application of this model to calculate downwelling radiation in all-sky conditions tends to result in larger errors for low altitude and humid climates. 

We propose to add the calibrated version of the Berdahl and Martin, Brunt and Idso models, identified as three models with higher accuracy compared to Clark-Allen and other existing popular models. The three models are calibrated using the radiation and meteorological measurements from the SURFRAD (Surface Radiation Budget Network) and ASOS (Automated Surface Observing System) operated by NOAA (National Oceanic and Atmospheric Administration). Currently seven SURFRAD stations are operating
in climatologically diverse regions over the contiguous United States including Bondville (in Illinois), Boulder (in Colorado), Desert Rock (in Nevada), Fort Peck (in Montana), Goodwin Creek (in Mississippi), Penn State University (in Pennsylvania) and Sioux Falls (in South Dakota) represent the climatological diversities.

Addresses new feature request https://github.com/NREL/EnergyPlus/issues/4999.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [x] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [x] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by nealkruis** [comment]
@mjwitte should this include Martin-Berdahl, too?

**Comment by xuanluo113** [comment]
@nealkruis @mjwitte Great suggestion. For the first round of literature review, we selected four models, namely *Berdahl & Martin*, *Brunt*, *Brutsaert* and *Prata* as the attached document introduced in detail. These four models all most cited by literature. However, literature also suggested the newly calibrated forms of the *Brunt* and * Idso* model, proposed in this NFP, have better accuracy over a larger spatial domain, as they are calibrated over larger measured datasets nationwide.

If investigated in ASHRAE Standard 140, we should probably include *Berdahl & Martin* as well. 

[~Sky Emittance Models.pdf](https://github.com/NREL/EnergyPlus/files/3747015/Sky.Emittance.Models.pdf)

**Comment by nrel-bot-2c** [comment]
@XuanLuoLBL @lgentile it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@XuanLuoLBL @lgentile it has been 28 days since this pull request was last updated.

File: doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex
Comment (mjwitte) [inline doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:73]: Minor, but I've usually seen this referenced as Martin Berdahl rather than Berdahl Martin. Please check recent literature to see what's the most common citation of this model. It may depend on whether one is referencing this 1983 LBNL report https://escholarship.org/content/qt71v36429/qt71v36429.pdf or the 1984 paper you've referenced. Are the equations the same in both references?

File: idd/Energy+.idd.in
Comment (mjwitte) [inline idd/Energy+.idd.in:1467]: ClarkAllen should be another choice here, and we can drop "Use" and "Model" from the key choices. So, this would be:
```
       \type choice
       \key ClarkAllen
       \key Brunt
       \key Idso
       \key BerdahlMartin (or MartinBerdahl???)
       \key ScheduleValue
       \key DifferenceScheduleDryBulbValue
       \key DifferenceScheduleDewPointValue
       \default ClarkAllen
```

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:170]: And I would add a parameter for ClarkAllen to keep this clear.

File: doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex
Comment (mjwitte) [inline doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex]: This sentence reads very differently from the original regarding horizontal infrared radiation. The original implies HorizIR is used in conjunction with the empirical model, but the new text implies HorizIR is a separate method (which doesn't have an option) so I'm confused by the revision.

Oh, I see from the code that it will use horizontal IR if present in the weather data, if one of the models is selected, but not if it's scheduled. That needs to be much more clear here. And should there be an option to always use the model even when horizontal IR is present?  Not sure which way I lean on that question.

File: doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex
Comment (mjwitte) [inline doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:100]: Applicable references from the NFP need to be added to the Engineering Ref section.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc]: Why all the extra gymnastics above? Seems that lines 3462:3466 could be eliminated and this line could be 
`if (!Environment(Envrn).WP_Type1==WP_ScheduleValue)`
Oh, but I see `WP_Type1` is the index of the WeatherProperty. So, seems like it would be more efficient to change `WP_Type1` to be `WP_SkyTempType` and set it when processing the WeatherProperty object? Then all of these ifs could be more direct.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc]: Same comment as above.

File: src/EnergyPlus/WeatherManager.hh
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.hh:111]: Do these all need to be `extern`? Looks like they're only used in WeatherManager.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:8048]: I'm not finding any code that reports the selected sky temperature model in the eio output. That *should* have happenend when WeatherProperty:* was added. Since this can be set per environment, it could be added to this eio output:
```
! <Environment>,Environment Name,Environment Type, Start Date, End Date, Start DayOfWeek, Duration {#days}, Source:Start DayOfWeek,  Use Daylight Saving, Use Holidays, Apply Weekend Holiday Rule,  Use Rain Values, Use Snow Values
```

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 Some initial comments.

File: src/EnergyPlus/WeatherManager.hh
Comment (xuanluo113) [reply src/EnergyPlus/WeatherManager.hh:111]: @mjwitte I "extern" them for the unit test only (otherwise they cannot be declared in the unit test file). Any better suggestions?

File: doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex
Comment (xuanluo113) [reply doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:73]: They use the same equation for sky emissivity.  

Emissivity of clear skies - Berdahl and Martin (https://doi.org/10.1016/0038-092X(84)90144-0) (cited 142) is the first to introduced the emissivity equation.  Martin and Berdahl (https://doi.org/10.1016/0038-092X(84)90162-2) (cited 158) introduced the sky radiant temperature equation from this emissivity estimation.

I still tend to call it Berdahl and Martin because *ESP-r* that currently uses this algorithm calls it Berdahl and Martin algorithm. (http://www.esru.strath.ac.uk/Programs/ESP-r_FAQ.htm)

Comment (mjwitte) [reply doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:73]: Ok

File: doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex
Comment (xuanluo113) [reply doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:100]: Done.

File: doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex
Comment (xuanluo113) [reply doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex]: The logic is re-illustrated.

File: src/EnergyPlus/WeatherManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/WeatherManager.cc]: I added a `SkyTempModel` field in the `Environment`. It's still required in `WeatherProperty` object for matching with Environments and reading in the schedules.

File: src/EnergyPlus/WeatherManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/WeatherManager.cc:8048]: Done.

**Comment by mjwitte** [comment]
@xuanluo113 Thanks for addressing comments. Is this ready for final review?

**Comment by xuanluo113** [comment]
> @xuanluo113 Thanks for addressing comments. Is this ready for final review?

@mjwitte Yes. Thanks, Mike.

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 The eio diffs are showing only that the `<Environment>` header line has the added field for Sky Temperature, but the `Environment` data line does not. It's possible I didn't fix the merge conflicts correctly, but looking back at earlier diffs, I see only the header line there as well.

I ran 5ZoneAutoDXVAV as-is (with no WeatherProperty object) and then added a WeatherProperty object with the defaults and got different results. These should be the same.

I ran 1ZoneUncontrolled_DDChanges as-is which has two WeatherProperty:SkyTemperature, one with a name, and one without. This crashes locally with a debug build. It is probably failing on the runperiod which doesn't run on CI.

**Comment by Myoldmopar** [comment]
EIO changes look to be in order now.  @mjwitte can you easily rerun the experiment to ensure a default weather property object gives the same results?  If not I am happy to do it.  CI otherwise looks happy.  I'm going to look over the code.

File: src/EnergyPlus/WeatherManager.hh
Comment (Myoldmopar) [reply src/EnergyPlus/WeatherManager.hh:111]: The ideal option would be to just use an enum class for options like this.  There are occasions where that is not feasible, such as if a variable of that enum type is used as an integer output variable.  But those are very few, and in most cases, changing to an enum can fix this.  I'm not pushing for that now, but in the future, that would be great.

File: tst/EnergyPlus/unit/WeatherManager.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/WeatherManager.unit.cc]: Should not be necessary to deallocate here.  I know it can be seen as a safeguard, but extra deallocations have a tendency to mask deallocations that should be in the appropriate `clear_state` functions.

File: src/EnergyPlus/WeatherManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WeatherManager.cc:842]: I would have preferred a vector here.

File: src/EnergyPlus/WeatherManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WeatherManager.cc]: Marking these as `const` would be best-practice

File: src/EnergyPlus/WeatherManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WeatherManager.cc:3778]: Nice compact, easy to test, function

**Review by Myoldmopar — APPROVED** [review]
I had some comments, but none urgent enough to necessitate pushing more commits unless there is other stuff to commit.

**Comment by mjwitte** [comment]
@Myoldmopar I'll retest today/tonight.

**Comment by mjwitte** [comment]
@xuanluo113 @Myoldmopar 1ZoneUncontrolled_DDChanges still crashes here in debug build on Windows. Will debug here an let you know what I find.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc]: @xuanluo113 1ZoneUncontrolled_DDChanges is failing here, because `OutHumRat` is uninitialized.  But looking at the signature for `CalcSkyEmissivity`, the last argument is `RelHum`, not `OutHumRat`.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [reply src/EnergyPlus/WeatherManager.cc]: Why assign all of these here when only some are used depending on the sky model type? Seems these should all be moved within the respective sky model where they are used.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:3753]: Why have `RelHum` as an argument when `DryBulb` and `DewPoint` are also here?

**Comment by xuanluo113** [comment]
[1Zone.zip](https://github.com/NREL/EnergyPlus/files/4201204/1Zone.zip)

Thanks, Mike and Edwin. Attached the local test cases and weather files for reference.

**Comment by mjwitte** [comment]
Confirmed that 5ZoneAutoDXVAV and 5ZoneAutoDXVAV with a default WeatherProperty:SkyTemperature object produced identical results.

File: src/EnergyPlus/WeatherManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/WeatherManager.cc:3753]: RelHum is used to calculate the partial water pressure directly. Otherwise, kinda have to convert it inside function.

File: src/EnergyPlus/WeatherManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/WeatherManager.cc]: That's also true @mjwitte.  Though PartialPress (and therefore SatPress) is used by two of them, and TDewC is implicitly used in two of them, so there would be some duplication.  I'm OK with leaving it like this or putting the initialization into each IF block.

**Comment by Myoldmopar** [comment]
@xuanluo113 you'll need to address the crash and the question from @mjwitte about the function signature.

`CalcSkyEmissivity` is declared as:

```
Real64 CalcSkyEmissivity(..., Real64 DryBulb, Real64 DewPoint, Real64 RelHum);
```

With relative humidity as the last argument.  You also have dry bulb and dew point so unless you are saying the barometric pressure is changing, you only need to pass in two psychrometric values to locate your state.  However, even with that variable there, you are calling it with humidity ratio:

```
ESky = CalcSkyEmissivity(..., DryBulb, TomorrowOutDewPointTemp(TS, Hour), OutHumRat);
```

And that's causing a crash because `OutHumRat` is uninitialized at that point.

Please address that as soon as possible and make sure to pull develop in right before you push your changes so we get clean CI results.

File: src/EnergyPlus/WeatherManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/WeatherManager.cc:3753]: > @xuanluo113 you'll need to address the crash and the question from @mjwitte about the function signature.
> 
> `CalcSkyEmissivity` is declared as:
> 
> ```
> Real64 CalcSkyEmissivity(..., Real64 DryBulb, Real64 DewPoint, Real64 RelHum);
> ```
> 
> With relative humidity as the last argument. You also have dry bulb and dew point so unless you are saying the barometric pressure is changing, you only need to pass in two psychrometric values to locate your state. However, even with that variable there, you are calling it with humidity ratio:
> 
> ```
> ESky = CalcSkyEmissivity(..., DryBulb, TomorrowOutDewPointTemp(TS, Hour), OutHumRat);
> ```
> 
> And that's causing a crash because `OutHumRat` is uninitialized at that point.
> 
> Please address that as soon as possible and make sure to pull develop in right before you push your changes so we get clean CI results.

@Myoldmopar @mjwitte Yes. I'm working on it now. But the calculation of partial vapor pressure requires relative humidity directly, while the other places requires dew point. If I only pass two psychrometric vars, I have to re-do the psychrometric conversion again inside function, which is a kind of duplicate calculation?

File: src/EnergyPlus/WeatherManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/WeatherManager.cc:842]: @Myoldmopar like a std::vector<string>? I thought this should be a const array.

**Comment by xuanluo113** [comment]
@mjwitte @Myoldmopar   I fixed the urgent ones and will wait for the CI. Three remaining issues:
1. `RelHum` is calculated in the caller. I'm not sure if it's better to (1) recalculate it again in the callee when used or (2) passing an extra argument sometimes not used.
2. I thought const array of string here could perform better than vector of strings.
3. Mike suggested adding another field in IDD `Use Weather File Horizontal IR`, which is implicitly `yes` right now. I only included the IDD change in this commit and is working on it while waiting for CI.

**Comment by mjwitte** [comment]
1. It's fine to pass `RelHum` rather than recalculate it.
2. I defer to @Myoldmopar 
3. Thanks.

**Comment by Myoldmopar** [comment]
1. Yeah, leave RelHum in there as well
2. I'm saying making it a const vector of strings.  We are moving away from Array1D soon, so adding new instances is counterproductive, that's all.  Honestly though, at this point, just leave it as-is.
3. :+1:

**Comment by xuanluo113** [comment]
@mjwitte @Myoldmopar The CI is clear and the eio is showing.

The small eso diff is due to the line I added the [here](https://github.com/NREL/EnergyPlus/blob/3496ad1a8282a7c251cf96c3463fbf8c56af7f97/src/EnergyPlus/WeatherManager.cc#L4547), which I think should be added.

**Comment by mjwitte** [comment]
@xuanluo113 I think I'm ok with the change to calculate `TomorrowOutRelHum` in the else block, but I'm puzzled trying to find where it was set previously such that the eso output in develop is rounded to a single decimal place, but this calculation is not. For example, what was 45.0 is now 44.9999999999.

**Review by mjwitte — COMMENTED** [review]
Tested various options (ClarkAllen, default, Brunt, ClarkAllen and Brunt with epw HorizIR=No), all looks good, saw changes in sky temp where expected, eio outputs were consistent with the input choices.

We're almost there.
1. Please add technical reference comments in `WeatherManager::CalcSkyEmissivity`
2. Add an entry for this in \src\Transition\OutputRulesFiles\OutputChanges9-2-0-to-9-3-0.md describing the new field at the end of the eio "Environment" lines.
3. The engineering ref needs some cleanup, exponent at the end of eq 5.5, and the subscript on the left side of eq 5.6. 
image

**Review by mjwitte — APPROVED** [review]
Confirmed requested changes. This is good to go.

**Pull request merged**
